pkg/infrared/protocol/login/serverbound_loginstart.go

```diff
@@ -13,35 +13,38 @@ const (
 )
 
 type ServerBoundLoginStart struct {
-	Name         protocol.String
-	HasPublicKey protocol.Boolean
+	Name protocol.String
 
 	// Added in 1.19; removed in 1.19.3
-	Timestamp protocol.Long
-	PublicKey protocol.ByteArray
-	Signature protocol.ByteArray
+	HasSignature protocol.Boolean
+	Timestamp    protocol.Long
+	PublicKey    protocol.ByteArray
+	Signature    protocol.ByteArray
 
 	// Added in 1.19
-	HasPlayerUUID protocol.Boolean
+	HasPlayerUUID protocol.Boolean // removed in 1.20.2
 	PlayerUUID    protocol.UUID
 }
 
 func (pk ServerBoundLoginStart) Marshal(packet *protocol.Packet, version protocol.Version) {
-	if version < protocol.Version_1_19 {
-		packet.Encode(
-			IDServerBoundLoginStart,
-			pk.Name,
-		)
-		return
-	}
-
 	fields := make([]protocol.FieldEncoder, 0, 7)
-	fields = append(fields, pk.Name, pk.HasPublicKey)
-	if pk.HasPublicKey {
-		fields = append(fields, pk.Timestamp, pk.PublicKey, pk.Signature)
-	}
-	fields = append(fields, pk.HasPlayerUUID)
-	if pk.HasPlayerUUID {
+	fields = append(fields, pk.Name)
+
+	switch {
+	case version >= protocol.Version_1_19 &&
+		version < protocol.Version_1_19_3:
+		fields = append(fields, pk.HasSignature)
+		if pk.HasSignature {
+			fields = append(fields, pk.Timestamp, pk.PublicKey, pk.Signature)
+		}
+		fallthrough
+	case version >= protocol.Version_1_19_3 &&
+		version < protocol.Version_1_20_2:
+		fields = append(fields, pk.HasPlayerUUID)
+		if pk.HasPlayerUUID {
+			fields = append(fields, pk.PlayerUUID)
+		}
+	case version >= protocol.Version_1_20_2:
```

Comment:
I don't think these condition will scale well, its also annoying that we need 2 switch cases with the same conditions here. 

How do you look at this?

Reply:
I think it's fine for now and I don't have any issue with it. Let's keep it as it is for now.

Reply:
Sure